workspaces/github-issues/plugins/search-backend-module-github-issues/src/collators/GithubIssuesCollatorFactory.ts

```diff
@@ -0,0 +1,234 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  LoggerService,
+  RootConfigService,
+} from '@backstage/backend-plugin-api';
+import { Config } from '@backstage/config';
+import {
+  DefaultGithubCredentialsProvider,
+  GithubCredentialsProvider,
+  ScmIntegrationRegistry,
+  ScmIntegrations,
+} from '@backstage/integration';
+import {
+  DocumentCollatorFactory,
+  IndexableDocument,
+} from '@backstage/plugin-search-common';
+import { graphql } from '@octokit/graphql';
+import { Discussion, Issue, Query } from '@octokit/graphql-schema';
+import { Readable } from 'stream';
+
+/**
+ * Extended IndexableDocument with GitHub specific properties
+ *
+ * @public
+ */
+export interface GithubIssuesDocument extends IndexableDocument {
+  title: string;
+  createdAt: string;
+  url: string;
+  repository: {
+    name: string;
+  };
+  author: string;
+  labels: string[];
+}
+
+export type GithubIssuesCollatorFactoryOptions = {
+  logger: LoggerService;
+  config: RootConfigService;
+  githubCredentialsProvider: GithubCredentialsProvider;
+  integrations: ScmIntegrationRegistry;
+};
+
+type GraphQL = typeof graphql;
+
+/**
+ * Search collator responsible for collecting GitHub issues and discussions to index.
+ *
+ * @public
+ */
+export class GithubIssuesCollatorFactory implements DocumentCollatorFactory {
+  public readonly type: string = 'github';
+
+  private readonly logger: LoggerService;
+  private readonly config: Config;
+
+  private readonly integrations: ScmIntegrationRegistry;
+  private readonly githubCredentialsProvider: GithubCredentialsProvider;
+
+  constructor(options: GithubIssuesCollatorFactoryOptions) {
+    this.integrations = options.integrations;
+    this.githubCredentialsProvider = options.githubCredentialsProvider;
+    this.logger = options.logger;
+    this.config = options.config;
+  }
+
+  static fromConfig(
+    config: Config,
+    options: {
+      logger: LoggerService;
+      githubCredentialsProvider?: GithubCredentialsProvider;
+    },
+  ) {
+    const integrations = ScmIntegrations.fromConfig(config);
+
+    return new GithubIssuesCollatorFactory({
+      config: config,
+      logger: options.logger,
+      githubCredentialsProvider:
+        options.githubCredentialsProvider ||
+        DefaultGithubCredentialsProvider.fromIntegrations(integrations),
+      integrations,
+    });
+  }
+
+  async getCollator() {
+    return Readable.from(this.execute());
+  }
+
+  async *execute(): AsyncGenerator<GithubIssuesDocument> {
+    this.logger.info('Starting collation of GitHub issues.');
+
+    const documents = await this.fetchGithubIssuesDocuments();
```

Comment:
if there are 1000 issues, this will cause 1000 issues to be loaded in memory.
We could lift some logic from the `fetchGithubIssuesDocuments`, making sure that `fetchGithubIssuesDocuments` returns just a batch of element. Something like:

```ts
  let nextPage = ...
  while(hasNext()) {
    const result = this.fetchGithubIssuesDocuments(nextPage);
    nextPage = result.nextPage;
    for (const document of result.documents) {
      yield { ...document, ...stuff };
    }
  }
```

so that the collator will be more memory efficient.

Reply:
Good point! Will change